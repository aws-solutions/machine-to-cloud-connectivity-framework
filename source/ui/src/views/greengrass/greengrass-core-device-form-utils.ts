// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import React from 'react';
import { CreatedBy, FormControlElement, GreengrassCoreDeviceDomIds, UserGreengrassCoreDevice } from '../../util/types';
import { validateGreengrassCoreDeviceName } from '../../util/utils';

type HandleValueChangeRequest = {
  domIds: GreengrassCoreDeviceDomIds;
  event: React.ChangeEvent<FormControlElement>;
  greengrassCoreDevices: UserGreengrassCoreDevice[];
  setCreatedBy: React.Dispatch<CreatedBy>;
  setGreengrassCoreDeviceName: React.Dispatch<string>;
  setGreengrassCoreDeviceStatus: React.Dispatch<string>;
  setIsGreengrassCoreDeviceNameValid: React.Dispatch<boolean>;
  setGreengrassCoreDeviceOSPlatform: React.Dispatch<string>;
  statusMap: Map<string, string>;
};

/**
 * Handles the value change.
 * @param props The handle value change function properties
 * @param props.domIds The DOM IDs
 * @param props.event The React change event
 * @param props.greengrassCoreDevices The Greengrass core devices
 * @param props.setCreatedBy Setting created by function
 * @param props.setGreengrassCoreDeviceName Setting Greengrass core device name function
 * @param props.setGreengrassCoreDeviceStatus Setting Greengrass core device status function
 * @param props.setIsGreengrassCoreDeviceNameValid Setting Greengrass core device name valid check function
 * @param props.statusMap The Greengrass core device status map
 */
export function handleValueChange(props: HandleValueChangeRequest): void {
  const {
    domIds,
    event,
    greengrassCoreDevices,
    setCreatedBy,
    setGreengrassCoreDeviceName,
    setGreengrassCoreDeviceStatus,
    setIsGreengrassCoreDeviceNameValid,
    setGreengrassCoreDeviceOSPlatform,
    statusMap
  } = props;
  const { id, value } = event.target;

  switch (id) {
    case domIds.category:
      setCreatedBy(value as CreatedBy);
      setGreengrassCoreDeviceName('');
      setGreengrassCoreDeviceStatus('');
      setIsGreengrassCoreDeviceNameValid(false);

      if (value === CreatedBy.USER && greengrassCoreDevices.length > 0) {
        setGreengrassCoreDeviceName(greengrassCoreDevices[0].coreDeviceThingName);
        setGreengrassCoreDeviceStatus(greengrassCoreDevices[0].status);
      }

      break;
    case domIds.greengrassCoreDeviceName:
      setGreengrassCoreDeviceName(value);

      if (event.target instanceof HTMLSelectElement) {
        setGreengrassCoreDeviceStatus(statusMap.get(value) || 'UNKNOWN');
      } else {
        setIsGreengrassCoreDeviceNameValid(validateGreengrassCoreDeviceName(value));
      }

      break;
    case domIds.osPlatform:
      setGreengrassCoreDeviceOSPlatform(value);
      break;
    default:
      break;
  }
}
