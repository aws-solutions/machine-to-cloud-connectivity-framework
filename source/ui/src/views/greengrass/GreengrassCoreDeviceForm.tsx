// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { I18n, Logger } from '@aws-amplify/core';
import React, { FormEvent, useState } from 'react';
import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Breadcrumb from 'react-bootstrap/Breadcrumb';
import Card from 'react-bootstrap/Card';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import { LinkContainer } from 'react-router-bootstrap';
import GreengrassCoreDeviceInputForm from './GreengrassCoreDeviceInputForm';
import { handleValueChange } from './greengrass-core-device-form-utils';
import MessageModal from '../../components/MessageModal';
import EmptyCol from '../../components/EmptyCol';
import EmptyRow from '../../components/EmptyRow';
import { LoadingSpinner } from '../../components/Loading';
import { UserGreengrassCoreDevicesHook } from '../../hooks/GreengrassCoreDeviceHook';
import { requestApi } from '../../util/apis';
import {
  CreatedBy,
  FormControlElement,
  GreengrassCoreDeviceControl,
  GreengrassCoreDeviceDomIds,
  GreengrassCoreDevicePostResponse,
  MessageModalType
} from '../../util/types';
import { getErrorMessage } from '../../util/utils';

const logger = new Logger('GreengrassCoreDeviceForm');
const domIds: GreengrassCoreDeviceDomIds = {
  category: 'category',
  greengrassCoreDeviceName: 'greengrassCoreDeviceName'
};

/**
 * Renders the Greengrass core device registration form.
 * @returns The Greengrass core device modal
 */
export default function GreengrassCoreDeviceForm(): React.JSX.Element {
  const [confirmMessage, setConfirmMessage] = useState<string>('');
  const [createdBy, setCreatedBy] = useState<CreatedBy>(CreatedBy.SYSTEM);
  const [greengrassCoreDeviceName, setGreengrassCoreDeviceName] = useState<string>('');
  const [greengrassCoreDeviceStatus, setGreengrassCoreDeviceStatus] = useState<string>('');
  const [hasConfirmActionError, setHasConfirmActionError] = useState<boolean>(false);
  const [isGreengrassCoreDeviceNameValid, setIsGreengrassCoreDeviceNameValid] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(false);
  const [message, setMessage] = useState<string | React.ReactNode>('');
  const [showConfirmMessageModal, setShowConfirmMessageModal] = useState<boolean>(false);
  const [showMessageModal, setShowMessageModal] = useState<boolean>(false);
  const { close, greengrassCoreDevices, statusMap } = UserGreengrassCoreDevicesHook({
    setHasConfirmActionError,
    setMessage,
    setShowMessageModal
  });
  const change = (event: React.ChangeEvent<FormControlElement>) => {
    handleValueChange({
      domIds,
      event,
      greengrassCoreDevices,
      setCreatedBy,
      setGreengrassCoreDeviceName,
      setGreengrassCoreDeviceStatus,
      setIsGreengrassCoreDeviceNameValid,
      statusMap
    });
  };

  /**
   * Handles the Greengrass core device registration. It shows the confirmation modal.
   * @param event The Greengrass core device handle form event
   */
  function handleGreengrassCoreDevice(event: FormEvent): void {
    event.preventDefault();
    setConfirmMessage(I18n.get('info.message.register.greengrass.core.device').replace('{}', greengrassCoreDeviceName));

    if (createdBy === CreatedBy.USER && greengrassCoreDeviceStatus !== 'HEALTHY') {
      setConfirmMessage(
        I18n.get('warning.message.unhealthy.greengrass.core.device').replace('{}', greengrassCoreDeviceName)
      );
    }

    setShowConfirmMessageModal(true);
  }

  /**
   * Registers a Greengrass core device.
   */
  async function registerGreengrassCoreDevice(): Promise<void> {
    setLoading(true);

    try {
      const response = (await requestApi({
        method: 'post',
        path: '/greengrass',
        options: {
          body: {
            name: greengrassCoreDeviceName,
            control: GreengrassCoreDeviceControl.CREATE,
            createdBy
          }
        }
      })) as GreengrassCoreDevicePostResponse;

      setMessage(<span>{response.message}</span>);
      setHasConfirmActionError(false);
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      logger.error(errorMessage);

      setShowConfirmMessageModal(false);
      setHasConfirmActionError(true);
      setMessage(
        <Alert variant="danger">
          {I18n.get('error.message.register.greengrass.core.device')}
          <br />
          {I18n.get('error')}: <code>{JSON.stringify(errorMessage)}</code>
          <br />
        </Alert>
      );
    }

    setLoading(false);
    setShowMessageModal(true);
  }

  return (
    <Container>
      <Breadcrumb>
        <LinkContainer to="/greengrass">
          <Breadcrumb.Item>{I18n.get('manage.greengrass.core.devices')}</Breadcrumb.Item>
        </LinkContainer>
        <Breadcrumb.Item active>{I18n.get('register.greengrass.core.device')}</Breadcrumb.Item>
      </Breadcrumb>
      <Card>
        <Card.Body>
          <Form onSubmit={handleGreengrassCoreDevice}>
            <Form.Group>
              <Form.Label>
                {I18n.get('category')} <span className="red-text">*</span>
              </Form.Label>
              <Form.Text muted>{I18n.get('description.category')}</Form.Text>
              <Form.Control
                data-testid="category-select"
                id={domIds.category}
                as="select"
                onChange={change}
                value={createdBy}>
                <option value={CreatedBy.SYSTEM}>{I18n.get('category.create.new.greengrass.core.device')}</option>
                <option value={CreatedBy.USER}>{I18n.get('category.bring.existing.greengrass.core.device')}</option>
              </Form.Control>
            </Form.Group>
            <GreengrassCoreDeviceInputForm
              createdBy={createdBy}
              domIds={domIds}
              greengrassCoreDeviceName={greengrassCoreDeviceName}
              greengrassCoreDevices={greengrassCoreDevices}
              handleValueChange={change}
              isGreengrassCoreDeviceNameValid={isGreengrassCoreDeviceNameValid}
            />
            <EmptyRow />
            <Row>
              <Col className="justify-content-center grid">
                <ButtonGroup>
                  <Button
                    id="register-greengrass-core-device"
                    variant="primary"
                    size="sm"
                    type="submit"
                    disabled={
                      loading ||
                      (createdBy === CreatedBy.SYSTEM && !isGreengrassCoreDeviceNameValid) ||
                      (createdBy === CreatedBy.USER && greengrassCoreDevices.length === 0)
                    }>
                    <LoadingSpinner loading={loading} />
                    {I18n.get('register')}
                  </Button>
                  <EmptyCol />
                  <Button
                    id="register-greengrass-core-device"
                    variant="secondary"
                    size="sm"
                    onClick={() => close()}
                    disabled={loading}>
                    {I18n.get('cancel')}
                  </Button>
                </ButtonGroup>
              </Col>
            </Row>
          </Form>
        </Card.Body>
      </Card>
      <MessageModal
        show={showConfirmMessageModal}
        hide={() => setShowConfirmMessageModal(false)}
        message={confirmMessage}
        modalType={MessageModalType.CONFIRM}
        confirmAction={() => registerGreengrassCoreDevice()}
      />
      <MessageModal
        show={showMessageModal}
        hide={() => setShowMessageModal(false)}
        message={message}
        modalType={MessageModalType.MESSAGE}
        confirmAction={hasConfirmActionError ? () => undefined : () => close()}
      />
    </Container>
  );
}
