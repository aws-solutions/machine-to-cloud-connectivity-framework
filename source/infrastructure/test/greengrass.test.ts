// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { CfnCondition, Stack } from 'aws-cdk-lib';
import { Bucket, CfnBucket } from 'aws-cdk-lib/aws-s3';
import { GreengrassConstruct } from '../lib/greengrass/greengrass';

test('M2C2 greengrass resource creation test', () => {
  const stack = new Stack();
  const s3LoggingBucket = new Bucket(stack, 'TestLoggingBucket');
  (<CfnBucket>s3LoggingBucket.node.defaultChild).overrideLogicalId('TestLoggingBucket');
  const greengrass = new GreengrassConstruct(stack, 'TestGreengrass', {
    kinesisStreamName: 'TestStream',
    s3LoggingBucket,
    solutionConfig: {
      solutionId: 'SO0070-Test',
      solutionVersion: 'v0.0.1-test',
      uuid: 'test-uuid'
    },
    timestreamKinesisStreamArn: 'arn:of:timestream:kinesis:stream',
    customResourcesFunctionArn: 'test-arn',
    shouldTeardownData: new CfnCondition(stack, 'TestCondition')
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(greengrass.greengrassResourceBucket).toBeDefined();
  expect(greengrass.iotCredentialsRoleArn).toBeDefined();
  expect(greengrass.iotPolicyName).toBeDefined();
  expect(greengrass.iotRoleAliasName).toBeDefined();
});
