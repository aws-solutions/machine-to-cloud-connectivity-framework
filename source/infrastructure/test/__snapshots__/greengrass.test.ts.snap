// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 greengrass resource creation test 1`] = `
Object {
  "Resources": Object {
    "TestGreengrassGreengrassIoTPolicy448C7F58": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W39",
              "reason": "The * resource for Greengrass actions is the minimum requirement.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "greengrass:GetComponentVersionArtifact",
                "greengrass:ResolveComponentCandidates",
                "greengrass:GetDeploymentConfiguration",
                "greengrass:ListThingGroupsForCoreDevice",
                "greengrass:PutCertificateAuthorities",
                "greengrass:VerifyClientDeviceIdentity",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "greengrass:VerifyClientDeviceIoTCertificateAssociation",
                "greengrass:Discover",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":thing/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "greengrass:GetConnectivityInfo",
                "greengrass:UpdateConnectivityInfo",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":greengrass:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":/greengrass/things/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IoT::Policy",
    },
    "TestGreengrassGreengrassResourceBucketB85DC1F7": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "AWS::StackName",
              },
              "-",
              Object {
                "Ref": "AWS::AccountId",
              },
              "-gg",
            ],
          ],
        },
        "LoggingConfiguration": Object {
          "LogFilePrefix": "m2c2/",
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestGreengrassGreengrassResourceBucketPolicy7016D610": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "TestGreengrassGreengrassResourceBucketB85DC1F7",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestGreengrassGreengrassResourceBucketB85DC1F7",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestGreengrassGreengrassResourceBucketB85DC1F7",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestGreengrassIoTCredentialsRoleC2F465F7": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "The * action/resource is required to control Greengrass resources fully. The * resource is required for iotsitewise:BatchPutAssetPropertyValue for Greengrass v2.",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "F3",
              "reason": "The * action is required to control Greengrass resources fully.",
            },
            Object {
              "id": "W11",
              "reason": "The * resource is required to control Greengrass resources fully and iotsitewise:BatchPutAssetPropertyValue actions for Greengrass v2.",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "credentials.iot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:DescribeLogStreams",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/greengrass/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CloudWatchPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "greengrass:*",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "GreengrassPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iot:GetThingShadow",
                    "iot:UpdateThingShadow",
                    "iot:DeleteThingShadow",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":iot:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":thing/GG_*",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":iot:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":thing/*-gcm",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":iot:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":thing/*-gda",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":iot:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":thing/*-gci",
                        ],
                      ],
                    },
                  ],
                },
                Object {
                  "Action": "iot:Publish",
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":iot:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":topic/m2c2/info/*",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":iot:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":topic/m2c2/error/*",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":iot:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":topic/m2c2/data/*",
                        ],
                      ],
                    },
                  ],
                },
                Object {
                  "Action": "iot:DescribeThing",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":thing/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "IoTPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "iotsitewise:BatchPutAssetPropertyValue",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "IoTSiteWisePolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kinesis:PutRecords",
                    "kinesis:PutRecord",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":kinesis:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":stream/TestStream",
                        ],
                      ],
                    },
                    "arn:of:timestream:kinesis:stream",
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KinesisPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "s3:GetBucketLocation",
                    "s3:GetObject",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "TestGreengrassGreengrassResourceBucketB85DC1F7",
                        "Arn",
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          Object {
                            "Fn::GetAtt": Array [
                              "TestGreengrassGreengrassResourceBucketB85DC1F7",
                              "Arn",
                            ],
                          },
                          "/*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "S3Policy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "secretsmanager:GetSecretValue",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":secretsmanager:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":secret:m2c2-*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "SecretsManagerPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestGreengrassIoTPolicyE612A24A": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W39",
              "reason": "The * resource for iot:Connect is required for the solution.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "iot:Connect",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "iot:GetThingShadow",
                "iot:UpdateThingShadow",
                "iot:DeleteThingShadow",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":thing/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "iot:Publish",
                "iot:Receive",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/$aws/things/*/greengrass/health/json",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/$aws/things/*/greengrassv2/health/json",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/m2c2/job/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/m2c2/info/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/m2c2/error/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/m2c2/data/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/$aws/things/*/jobs/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/$aws/things/*/shadow/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/$aws/sitewise/gateways/*/diagnostics",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topic/$aws/sitewise/things/*/connectors/*/configuration/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "iot:Subscribe",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topicfilter/m2c2/job/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topicfilter/$aws/things/*/jobs/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topicfilter/$aws/things/*/shadow/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topicfilter/$aws/sitewise/gateways/*/diagnostics",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iot:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":topicfilter/$aws/sitewise/things/*/connectors/*/configuration/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "iot:AssumeRoleWithCertificate",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":rolealias/m2c2-role-alias-test-uuid",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IoT::Policy",
    },
    "TestGreengrassTeardownGreengrassResourcesBucket94AE96EB": Object {
      "Condition": "TestCondition",
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": Object {
          "Ref": "TestGreengrassGreengrassResourceBucketB85DC1F7",
        },
        "Resource": "DeleteS3Bucket",
        "ServiceToken": "test-arn",
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "TestLoggingBucket": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
